File: back-bardak/src/main/java/kz/bardak/game/ws/Envelope.java
package kz.bardak.game.ws;

import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.databind.JsonNode;

import java.time.Instant;

/**
 * Конверт WebSocket-сообщения, одинаковый в обе стороны.
 * Описан в planning/05-api-contracts.md.
 *
 * @param v        версия протокола; несовпадение → ERROR PROTOCOL_VERSION_UNSUPPORTED
 * @param id       идентификатор сообщения: у команд — для идемпотентности и ack
 * @param type     тип команды (client→server) или события (server→client)
 * @param tableId  стол, к которому относится сообщение; по нему идёт маршрутизация
 * @param seq      только у событий: сквозной номер по матчу
 * @param ts       отметка времени отправителя, epoch millis
 * @param payload  полезная нагрузка, форма зависит от type
 */
@JsonInclude(JsonInclude.Include.NON_NULL)
public record Envelope(
        Integer v,
        String id,
        String type,
        String tableId,
        Integer seq,
        Long ts,
        JsonNode payload
) {

    /** Текущая версия протокола. Ломающее изменение → +1. */
    public static final int PROTOCOL_VERSION = 1;

    /** Событие в ответ на команду: без seq, потому что вне матча нумерации ещё нет. */
    public static Envelope event(String type, String commandId, String tableId, JsonNode payload) {
        return new Envelope(PROTOCOL_VERSION, commandId, type, tableId, null,
                Instant.now().toEpochMilli(), payload);
    }
}

File: back-go/internal/transport/ws/envelope.go
// Package ws — WebSocket: конверт протокола, жизненный цикл соединений и рассылка.
package ws

import (
	"encoding/json"
	"time"
)

// ProtocolVersion — текущая версия протокола. Ломающее изменение → +1.
const ProtocolVersion = 1

// Envelope — конверт сообщения, ОДИНАКОВЫЙ в обе стороны.
//
// ⚠️ Правила пропуска полей повторяют Java и различаются по типам:
//   - `v` и `ts` у сервера есть всегда — обычные значения;
//   - `id`, `tableId`, `seq` бывают пустыми и выпадают из JSON целиком, а не приходят
//     как null — поэтому указатели с omitempty;
//   - `payload` — сырой JSON: у PONG его нет вовсе, а у ECHO он бывает явным null.
//
// ⚠️ Числа 0 и false НЕ выпадают: в Java стоит NON_NULL, а не NON_DEFAULT. Поставить
// omitempty на int означало бы потерять `mySeat: 0` и `passed: false`.
type Envelope struct {
	V       int             `json:"v"`
	ID      *string         `json:"id,omitempty"`
	Type    string          `json:"type"`
	TableID *string         `json:"tableId,omitempty"`
	Seq     *int            `json:"seq,omitempty"`
	TS      int64           `json:"ts"`
	Payload json.RawMessage `json:"payload,omitempty"`
}

// Event собирает событие сервера.
//
// `commandID` пуст у всех событий, кроме ERROR, PONG и ECHO, где он равен идентификатору
// вызвавшей команды.
func Event(eventType string, commandID, tableID *string, payload any) Envelope {
	envelope := Envelope{
		V:       ProtocolVersion,
		ID:      commandID,
		Type:    eventType,
		TableID: tableID,
		TS:      time.Now().UnixMilli(),
	}
	if payload != nil {
		if raw, err := json.Marshal(payload); err == nil {
			envelope.Payload = raw
		}
	}
	return envelope
}

// GameEvent — событие матча со сквозным номером.
//
// ⭐ `seq` есть ТОЛЬКО у игровых событий. У снимка состояния, у событий лобби и у всех
// служебных его нет — клиент отличает их именно по этому.
func GameEvent(eventType string, tableID string, seq int, payload any) Envelope {
	envelope := Event(eventType, nil, &tableID, payload)
	envelope.Seq = &seq
	return envelope
}

// ErrorEvent — отказ в ответ на команду.
func ErrorEvent(commandID, tableID *string, code, message string) Envelope {
	return Event("ERROR", commandID, tableID, map[string]string{
		"code":    code,
		"message": message,
	})
}

// ParseEnvelope разбирает входящее сообщение.
//
// ⚠️ Неизвестные поля молча игнорируются — как у Jackson с FAIL_ON_UNKNOWN_PROPERTIES=false.
// Клиент вправе прислать лишнее, и это не повод рвать соединение.
func ParseEnvelope(raw []byte) (Envelope, error) {
	var envelope Envelope
	if err := json.Unmarshal(raw, &envelope); err != nil {
		return Envelope{}, err
	}
	return envelope, nil
}

// StringPtr — короткая обёртка: указатели на строки нужны здесь на каждом шагу.
func StringPtr(value string) *string {
	if value == "" {
		return nil
	}
	return &value
}
